src/base/bittorrent/torrentparamrules.cpp

```diff
@@ -0,0 +1,722 @@
+        case ConditionType::AnyOf:
+            return u"AnyOf"_s;
+        case ConditionType::Not:
+            return u"Not"_s;
+        case ConditionType::AnyFilePathRegex:
+            return u"AnyFilePathRegex"_s;
+        case ConditionType::AnyTrackerUrlRegex:
+            return u"AnyTrackerUrlRegex"_s;
+        }
+        return u"UnrecognizedConditionType"_s;  // Should never be reached.
+    }
+
+    // Values used in JSON. Changing existing strings may break compatibility.
+    QString ModifierTypeStr(ModifierType t)
+    {
+        switch (t) {
+        case ModifierType::Compound:
+            return u"Compound"_s;
+        case ModifierType::AddTag:
+            return u"AddTag"_s;
+        case ModifierType::SetCategory:
+            return u"SetCategory"_s;
+        case ModifierType::SetSavePath:
+            return u"SetSavePath"_s;
+        }
+        return u"UnrecognizedModifierType"_s;  // Should never be reached.
+    }
+
+    std::optional<ConditionType> decodeConditionType(const QString &type)
+    {
+        using CT = ConditionType;
+        static const auto *const map = [] {
+                                           auto *m = new QMap<QString, CT>();
+                                           (*m)[conditionTypeStr(CT::AllOf)] = CT::AllOf;
+                                           (*m)[conditionTypeStr(CT::AnyOf)] = CT::AnyOf;
+                                           (*m)[conditionTypeStr(CT::Not)] = CT::Not;
+                                           (*m)[conditionTypeStr(CT::AnyFilePathRegex)] = CT::AnyFilePathRegex;
+                                           (*m)[conditionTypeStr(CT::AnyTrackerUrlRegex)] = CT::AnyTrackerUrlRegex;
+                                           return m;
+                                       }();
+        if (auto it = map->find(type); it != map->end())
+            return *it;
+        return std::nullopt;
+    }
+
+    std::optional<ModifierType> decodeModifierType(const QString &type)
+    {
+        using MT = ModifierType;
+        static const auto *const map = [] {
+                                           auto *m = new QMap<QString, MT>();
+                                           (*m)[ModifierTypeStr(MT::Compound)] = MT::Compound;
+                                           (*m)[ModifierTypeStr(MT::AddTag)] = MT::AddTag;
+                                           (*m)[ModifierTypeStr(MT::SetCategory)] = MT::SetCategory;
+                                           (*m)[ModifierTypeStr(MT::SetSavePath)] = MT::SetSavePath;
+                                           return m;
+                                       }();
+        if (auto it = map->find(type); it != map->end())
+            return *it;
+        return std::nullopt;
+    }
+
+    bool checkRequiredField(const QJsonObject &obj, const QString &key, QJsonValue::Type type)
+    {
+        auto field = obj.find(key);
+        if ((field == obj.end()) || (field.value().type() != type)) {
```

Comment:
Please consider qBittorrent Coding Guidelines and fix your code to improve reviewing process.

Reply:
Ack